operator/pkg/ingress/ingress_class.go

```diff
@@ -4,172 +4,181 @@
-	oldIngressClass *slim_networkingv1.IngressClass
-	newIngressClass *slim_networkingv1.IngressClass
-}
-
-type ingressClassDeletedEvent struct {
-	ingressClass *slim_networkingv1.IngressClass
-}
-
-type ingressClassManager struct {
-	informer cache.Controller
-	store    cache.Store
-
-	queue      workqueue.RateLimitingInterface
-	maxRetries int
-
-	ingressQueue workqueue.RateLimitingInterface
-}
-
-// type used to signal changes to the ingress controller queue
 type ciliumIngressClassUpdatedEvent struct {
-	ingressClass *slim_networkingv1.IngressClass
+	isDefault bool
+	changed   bool
 }
 
 type ciliumIngressClassDeletedEvent struct {
-	ingressClass *slim_networkingv1.IngressClass
+	wasDefault bool
 }
 
-func newIngressClassManager(clientset k8sClient.Clientset, ingressQueue workqueue.RateLimitingInterface, maxRetries int) (*ingressClassManager, error) {
+type ingressClassManager struct {
+	isDefaultIngressClass atomic.Bool
+	synced                atomic.Bool
+	ingressClasses        resource.Resource[*slim_networkingv1.IngressClass]
+	queue                 workqueue.RateLimitingInterface
+	events                <-chan resource.Event[*slim_networkingv1.IngressClass]
+}
+
+func newIngressClassManager(
+	ctx context.Context,
+	queue workqueue.RateLimitingInterface,
+	ingressClasses resource.Resource[*slim_networkingv1.IngressClass],
+) (*ingressClassManager, error) {
 	manager := &ingressClassManager{
-		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		ingressQueue: ingressQueue,
-		maxRetries:   maxRetries,
+		isDefaultIngressClass: atomic.Bool{},
+		synced:                atomic.Bool{},
+		ingressClasses:        ingressClasses,
+		events:                ingressClasses.Events(context.Background()),
+		queue:                 queue,
 	}
 
-	manager.store, manager.informer = informer.NewInformer(
-		utils.ListerWatcherFromTyped[*slim_networkingv1.IngressClassList](clientset.Slim().NetworkingV1().IngressClasses()),
-		&slim_networkingv1.IngressClass{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc:    manager.handleAddIngressClass,
-			UpdateFunc: manager.handleUpdateIngressClass,
-			DeleteFunc: manager.handleDeleteIngressClass,
-		},
-		nil,
-	)
-
-	go manager.informer.Run(wait.NeverStop)
-	if !cache.WaitForCacheSync(wait.NeverStop, manager.informer.HasSynced) {
-		return manager, fmt.Errorf("unable to sync service")
-	}
-	log.WithField("existing-ingressclasses", manager.store.ListKeys()).Debug("ingress classes synced")
+	manager.isDefaultIngressClass.Store(false)
+	manager.synced.Store(false)
+
 	return manager, nil
 }
 
-// Run kicks off the control loop
-func (i *ingressClassManager) Run() {
-	for i.processEvent() {
+func (i *ingressClassManager) IsDefault() bool {
+	return i.isDefaultIngressClass.Load()
+}
+
+func (i *ingressClassManager) processEvents(ctx context.Context, initialSync bool) error {
+	for {
+		select {
+		case event := <-i.events:
+			if event.Kind == resource.Sync {
```

Comment:
why handling the kind `sync` in it's own if statement instead of including in the existing switch-statement? i think it would increase the readability of this function.